File: api/schemas/history.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

HistoryStatus = Literal["queued", "running", "success", "failed", "cancelled"]
MessageRole = Literal["user", "assistant", "system", "tool"]
AssetType = Literal["video", "image", "audio", "subtitle", "workflow", "preview", "mask"]


class SessionSummary(BaseModel):
    id: str
    title: str
    user_id: Optional[str] = None
    project_type: str = "video_agent"
    status: HistoryStatus
    job_count: int = 1
    latest_job_id: Optional[str] = None
    thumbnail_url: Optional[str] = None
    video_url: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class HistoryMessage(BaseModel):
    id: str
    session_id: str
    role: MessageRole
    content: Dict[str, Any]
    created_at: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class GenerationJob(BaseModel):
    id: str
    session_id: str
    message_id: Optional[str] = None
    status: HistoryStatus
    progress: int = 0
    provider: str = "local"
    external_job_id: Optional[str] = None
    prompt: str = ""
    negative_prompt: Optional[str] = None
    model_name: Optional[str] = None
    workflow_id: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    duration: Optional[float] = None
    fps: Optional[int] = None
    seed: Optional[int] = None
    error_message: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    completed_at: Optional[str] = None
    params: Dict[str, Any] = Field(default_factory=dict)
    result: Dict[str, Any] = Field(default_factory=dict)


class HistoryAsset(BaseModel):
    id: str
    job_id: str
    session_id: str
    asset_type: AssetType
    url: Optional[str] = None
    local_path: Optional[str] = None
    thumbnail_url: Optional[str] = None
    filename: Optional[str] = None
    mime_type: Optional[str] = None
    size_bytes: Optional[int] = None
    width: Optional[int] = None
    height: Optional[int] = None
    duration: Optional[float] = None
    created_at: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class WorkflowSnapshot(BaseModel):
    id: str
    job_id: str
    session_id: str
    workflow_name: str
    workflow_json: Dict[str, Any] = Field(default_factory=dict)
    ui_json: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None


class SessionListResponse(BaseModel):
    sessions: List[SessionSummary]
    total: int


class SessionDetail(BaseModel):
    session: SessionSummary
    messages: List[HistoryMessage]
    generation_jobs: List[GenerationJob]
    assets: List[HistoryAsset]
    workflow_snapshots: List[WorkflowSnapshot]

File: frontend/src/types/history.ts
export type HistoryStatus = "queued" | "running" | "success" | "failed" | "cancelled";
export type MessageRole = "user" | "assistant" | "system" | "tool";
export type AssetType = "video" | "image" | "audio" | "subtitle" | "workflow" | "preview" | "mask";

export interface SessionSummary {
  id: string;
  title: string;
  user_id?: string | null;
  project_type: string;
  status: HistoryStatus;
  job_count: number;
  latest_job_id?: string | null;
  thumbnail_url?: string | null;
  video_url?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  metadata: Record<string, unknown>;
}

export interface HistoryMessage {
  id: string;
  session_id: string;
  role: MessageRole;
  content: {
    text?: string;
    attachments?: unknown[];
    intent?: string;
    agentSummary?: string;
    scenes?: HistoryScene[];
    [key: string]: unknown;
  };
  created_at?: string | null;
  metadata: Record<string, unknown>;
}

export interface HistoryScene {
  index: number;
  narration: string;
  image_prompt: string;
  visual_description?: string;
  media_prompt?: string;
  duration?: number;
  estimated_duration?: number;
  asset_type?: "image" | "video";
  media_type?: "image" | "video";
  status?: "draft" | "queued" | "running" | "completed" | "failed";
  research_job_id?: string | null;
  subject_id?: string | null;
  claim_ids?: string[];
  visual_fact_ids?: string[];
  field_provenance?: Record<string, import("@/types/api").FieldProvenance>;
  fallback_level?: import("@/types/api").FallbackLevel;
  verification_status?: import("@/types/api").VerificationStatus;
  negative_constraints?: string[];
  warnings?: string[];
  reference_asset_ids?: string[];
}

export interface GenerationJob {
  id: string;
  session_id: string;
  message_id?: string | null;
  status: HistoryStatus;
  progress: number;
  provider: string;
  external_job_id?: string | null;
  prompt: string;
  negative_prompt?: string | null;
  model_name?: string | null;
  workflow_id?: string | null;
  width?: number | null;
  height?: number | null;
  duration?: number | null;
  fps?: number | null;
  seed?: number | null;
  error_message?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  completed_at?: string | null;
  params: Record<string, unknown>;
  result: Record<string, unknown>;
}

export interface HistoryAsset {
  id: string;
  job_id: string;
  session_id: string;
  asset_type: AssetType;
  url?: string | null;
  local_path?: string | null;
  thumbnail_url?: string | null;
  filename?: string | null;
  mime_type?: string | null;
  size_bytes?: number | null;
  width?: number | null;
  height?: number | null;
  duration?: number | null;
  created_at?: string | null;
  metadata: Record<string, unknown>;
}

export interface WorkflowSnapshot {
  id: string;
  job_id: string;
  session_id: string;
  workflow_name: string;
  workflow_json: Record<string, unknown>;
  ui_json: Record<string, unknown>;
  created_at?: string | null;
}

export interface SessionListResponse {
  sessions: SessionSummary[];
  total: number;
}

export interface SessionDetail {
  session: SessionSummary;
  messages: HistoryMessage[];
  generation_jobs: GenerationJob[];
  assets: HistoryAsset[];
  workflow_snapshots: WorkflowSnapshot[];
}

export interface RetryJobResponse {
  params: Record<string, unknown>;
}
